nixos/forgejo: changelog and migration instructions

bendlas opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bendlas committed:
nixos/forgejo: changelog and migration instructions

bendlas commented:
cc @schmittlauch as per https://github.com/NixOS/nixpkgs/pull/248310#issuecomment-1741817292

schmittlauch commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -139,6 +139,8 @@
 
 - [c2FmZQ](https://github.com/c2FmZQ/c2FmZQ/), an application that can securely encrypt, store, and share files, including but not limited to pictures and videos. Available as [services.c2fmzq-server](#opt-services.c2fmzq-server.enable).
 
+- [forgejo](https://forgejo.org/), a git forge. Previously deployed as a drop-in to gitea. Available as [services.forgejo](#opt-services.forgejo.enable). See migration instructions in [manual](https://nixos.org/manual/nixos/stable#module-forgejo)
```
```suggestion
- [forgejo](https://forgejo.org/), a git forge. Previously deployed as a drop-in replacement package in the gitea module. Available as [services.forgejo](#opt-services.forgejo.enable). See migration instructions in [manual](https://nixos.org/manual/nixos/stable#module-forgejo)
```

schmittlauch replied:
Possibly also add that the migration from the drop-in package with gitea module to the new forgejo module does not have to happen immediately. This can be relevant for automated rollouts of unattended upgrades, where manual migration steps shall be avoided.

bendlas replied:
:heavy_check_mark:

schmittlauch reviewed (changes requested):
I haven't tried the migration steps myself so far (as I do not intend to upgrade my stable forgejo instance yet), but they look plausible to me. I added some possible improvements as comments.

As issue discussions often tend to focus more on complaining and nitpicking of negative things, I explicitly want to thank you for providing the upgrade instructions lacking in the last PR. Things are going well. :hugs:

bendlas committed:
nixos/forgejo/docs: clarify sentence

Co-authored-by: Trolli Schmittlauch <schmittlauch@users.noreply.github.com>

bendlas committed:
nixos/forgejo/docs: document migration via gitea impersonation

bendlas committed:
nixos/forgejo/docs: note about url change on migration

bendlas commented:
> I do not intend to upgrade my stable forgejo instance yet

Maybe I could interest you in verifying the instructions for running the forgejo module in gitea mode, that I added? Because I don't think I've got an instance with gitea module left :see_no_evil:

bendlas committed:
nixos/forgejo/docs: note about migration (non-)requirement

bendlas committed:
nixos/forgejo/docs: header ids

schmittlauch commented:
You might be lucky, I see what me and my btrfs snapshots can do this week ;)

emilylange commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -139,6 +139,8 @@
 
 - [c2FmZQ](https://github.com/c2FmZQ/c2FmZQ/), an application that can securely encrypt, store, and share files, including but not limited to pictures and videos. Available as [services.c2fmzq-server](#opt-services.c2fmzq-server.enable).
 
+- [forgejo](https://forgejo.org/), a git forge. Previously deployed as a drop-in replacement package in the gitea module. Available as [services.forgejo](#opt-services.forgejo.enable). See migration instructions in [manual](https://nixos.org/manual/nixos/stable#module-forgejo)
```
```suggestion
- [forgejo](https://forgejo.org/), a git forge. Previously deployed as a drop-in replacement package in the [gitea module](#opt-services.gitea.package). Available as [services.forgejo](#opt-services.forgejo.enable). See migration instructions in the [NixOS manual](#module-forgejo) on how to migrate your forgejo instance using [`services.gitea.package = pkgs.forgejo`](#opt-services.gitea.package) to [`services.forgejo`](#opt-services.forgejo.enable).
```

emilylange commented on nixos/doc/manual/release-notes/rl-2311.section.md:
Note https://github.com/NixOS/nixpkgs/blob/a2b6873f0b7f1468fa4a3e759c426b28dae8908d/nixos/doc/manual/release-notes/rl-2311.section.md?plain=1#L53

bendlas replied:
:heavy_check_mark:

emilylange commented on nixos/modules/services/misc/forgejo.md:
WIP, but no idea how to improve it further.
I tested the "full-migration" a while ago and fixed some typos in [your original comment](https://github.com/NixOS/nixpkgs/pull/248310#issuecomment-1676480490), btw @bendlas.

```diff
diff --git a/nixos/modules/services/misc/forgejo.md b/nixos/modules/services/misc/forgejo.md
index 366fbd008aee..6a3407382085 100644
--- a/nixos/modules/services/misc/forgejo.md
+++ b/nixos/modules/services/misc/forgejo.md
@@ -4,23 +4,37 @@ Forgejo is a soft-fork of gitea, with strong community focus, as well
 as on self-hosting and federation. [Codeberg](https://codeberg.org) is
 deployed from it.
 
-See [upstream docs](https://forgejo.org/docs/latest/)
+See [upstream docs](https://forgejo.org/docs/latest/).
+
+The method of choice for running forgejo is using [`services.forgejo`](#opt-services.forgejo.enable).
+
+::: {.warning}
+Running forgejo using `services.gitea.package = pkgs.forgejo` is no longer
+recommended.
+If you experience issues with your instance using `services.gitea`,
+**DO NOT** report them to the `services.gitea` module maintainers.
+**DO** report them to the `services.forgejo` module maintainers instead.
+:::
 
 ## Migration from Gitea {#module-forgejo-migration-gitea}
 
-The method of choice for running forgejo used to be with the
-`services.gitea` module. For migrating such a setup to the
-`services.forgejo` module:
+::: {.note}
+Migrating is, while not strictly necessary at this point, highly recommended.
+Both modules and projects are likely to divide further with each release.
+Which might lead to an even more involved migration.
+:::
 
-**NOTE:** migration is not strictly necessary at this point, since the
-gitea - drop-in method is expected to continue working for the time
-being.
+### Full-Migration {#module-forgejo-migration-gitea-default}
 
-### for default options {#module-forgejo-migration-gitea-default}
+This will migrate the state directory (data), rename and chown the database and
+delete the gitea user.
 
-for PostgreSQL, adapt accordingly on other DBs
+::: {.note}
+This will also change the git remote ssh-url user from `gitea@` to `forgejo@`,
+when using the host's openssh server (default) instead of the integrated one.
+:::
 
-**NOTE:** this will also change git url user from `gitea@` to `forgejo@`
+Instructions for PostgreSQL (default). Adapt accordingly for other databases:
 
 ```sh
 systemctl stop gitea
@@ -30,18 +44,21 @@ runuser -u postgres -- psql -c '
   ALTER DATABASE gitea RENAME TO forgejo;
 '
 nixos-rebuild switch
+systemctl stop forgejo
 chown -R forgejo:forgejo /var/lib/forgejo
 systemctl restart forgejo
 ```
 
-### impersonating gitea {#module-forgejo-migration-gitea-impersonate}
+### Alternatively, keeping the gitea user {#module-forgejo-migration-gitea-impersonate}
 
-Alternatively, instead of renaming the database, the state folder and
-changing the user, the forgejo module can be set up to directly work
-with gitea state. Be careful to disable `services.gitea`, when doing
-this.
+Alternatively, instead of renaming the database, copying the state folder and
+changing the user, the forgejo module can be set up to re-use the old storage
+locations and database, instead of having to copy or rename them.
+Make sure to disable `services.gitea`, when doing this.
 
 ```nix
+services.gitea.enable = false;
+
 services.forgejo = {
   enable = true;
   user = "gitea";
@@ -50,4 +67,13 @@ services.forgejo = {
   database.name = "gitea";
   database.user = "gitea";
 };
+
+users.users,gitea = {
+  home = "/var/lib/gitea";
+  useDefaultShell = true;
+  group = "gitea";
+  isSystemUser = true;
+};
+
+users.groups.gitea = {};
 ```
```

bendlas replied:
Applied. Thank you very much!

emilylange reviewed (dismissed):
You can test the change by building the manual yourself:
```
nix-build nixos/release.nix -A manual.x86_64-linux
```

bendlas committed:
nixos/forgejo/docs: clarify release notes entry

Co-authored-by: Emily <git@emilylange.de>

bendlas committed:
nixos/forgejo/docs: improve manual entry

Co-authored-by: Emily <git@emilylange.de>

bendlas committed:
nixos/forgejo/docs: move changelog line to the middle of the section

as noted <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->

bendlas commented:
Branch - off in 3 days, so last chance to review this ...

bjornfor commented:
I think 8 of 9 commits is a single logical change -- please squash?

RaitoBezarius commented:
@bendlas Please don't be so much in a hurry for such changes, we can always backport them, and they are not critical for branch-off. Leave the time to review them.

bendlas commented:
ok, sorry

schmittlauch reviewed (commented):
I successfully tested the *migration by gitea-impersonation* approach, it works like a charm.
As I made the error of running gitea on mariadb and renaming databases there is kind of a pain, I unfortunately could not verify the full migration instructions.

Apart from that I discovered a typo. I can of course submit a follow-up PR for this, but preferred to keep the review feedback tied to the original PR.

schmittlauch commented on nixos/modules/services/misc/forgejo.md:
```diff
@@ -0,0 +1,79 @@
+  database.user = "gitea";
+};
+
+users.users,gitea = {
```
Ah, I see this has already been fixed in 8c441fc2a765b.

bendlas commented on nixos/modules/services/misc/forgejo.md:
```diff
@@ -0,0 +1,79 @@
+
+::: {.note}
+Migrating is, while not strictly necessary at this point, highly recommended.
+Both modules and projects are likely to divide further with each release.
```
please do create a new PR

schmittlauch replied:
I just did: #270256
